[23.2] Fix creating new erronous workflow can produce multiple empty workflows

ahmedhamidawan opened the pull request:
Use `POST api/workflows` to create and save-as workflows, instead of the `workflows/create` and `workflows/save_workflow_as` controllers. This prevents the bug of creating (or saving as) workflows with errors and getting no prompts.
On the front-end, handle errors better and show them in editor error modal.

Fixes https://github.com/galaxyproject/galaxy/issues/17405
Fixes https://github.com/galaxyproject/galaxy/issues/17130

### Before:
https://github.com/galaxyproject/galaxy/assets/78516064/ebdf71f8-da6f-4e8f-9e0c-08e10044d930

### After:
https://github.com/galaxyproject/galaxy/assets/78516064/72b2feda-bb34-4cc7-a6af-9d8fd1257c1c


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -432,15 +428,29 @@ def save_workflow_as(
+                session.commit()
+
+            if missing_tools_return:
+                return missing_tools_return
```
Why did you make that change, it doesn't look related or correct.

ahmedhamidawan replied:
I moved the Persist after error checking 
for Duplicate labels because if the Persist is before the update function call, we create a new empty workflow, then add the steps, and then tell the user that their workflow has errors. This way, the new workflow is only created if we don't have the Duplicate label error.

However, for missing tools, we used to create the new workflow without the steps and then tell the user that their workflow has missing tools. So, I created the `missing_tools_return` variable to make sure we don't return without Persisting in the case of the missing tools error.

_I can make this treat the Duplicate label error the same way as we treat the Missing tools error if we're ok with that?_

mvdbeek replied:
I would prefer that we focus on the bug before refactoring things. I think that's not a clearcut improvement.

ahmedhamidawan replied:
But in order to fix this bug, we need to catch this exception

mvdbeek replied:
No, you need to use the API. This used to work fine on 23.1

ahmedhamidawan replied:
Ok, I'll look at what we did in 23.1, and try to use the api instead of the controller here.

mvdbeek replied:
What you want to do is replace `save_workflow_as` with a PUT to `/api/workflows/{id}` or a POST to `/api/workflows` (depending on whether the workflow has been saved before).

ahmedhamidawan replied:
Done, used `POST api/workflows` instead for both creating and saving-as workflows. It works and the errors are correctly handled at the client as well.

mvdbeek commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -432,15 +428,29 @@ def save_workflow_as(
                     workflow_update_options,
                 )
+            except exceptions.DuplicatedIdentifierException as e:
+                return self.message_exception(trans, e.err_msg, False)
```
This is only necessary because you're using the controller method, which you don't need to use. The API handles this correctly, `DuplicatedIdentifierException` derives from MessageException.

ahmedhamidawan replied:
Ah ok, now I understand why the exception wasn't being caught at all

ahmedhamidawan committed:
use `api/workflows` endpoint instead of `save_workflow_as` controller

ahmedhamidawan committed:
fix workflow editor creater bugs

Prevent the "expected Object got Array" error message.
Ensure `CreatorEditor` is loaded with existing creator(s).

ahmedhamidawan committed:
fix request handling on client for WorkflowEditor create/save as

ahmedhamidawan committed:
use POST api/workflows instead of `save_workflow_as` controller

mvdbeek commented on client/src/components/Workflow/Editor/Index.vue:
```diff
@@ -532,30 +536,27 @@ export default {
+            if (!this.saveAsName && !this.nameValidate()) {
+                return;
+            }
+            const rename_name = this.saveAsName ?? `SavedAs_${this.name}`;
```
The `SavedAs` case can't actually happen because of the check on line 540, or am I misreading this ?

ahmedhamidawan replied:
You are right, i messed this up refactoring the Save as function, thanks for pointing this out

ahmedhamidawan replied:
Actually, no, that would be the case if line 540 had an OR

**line 540:**
- if user doesn't provide a _Save As name_ in the save as modal, 
- AND the user doesn't have a `this.name` (set in the Editor right panel name input field)

_return_

**Now in line 543:**
- it is still possible that the user didn't provide a _Save As name_, but they might have still had a `this.name`, so save the workflow as "SaveAs_OGWf name"

mvdbeek reviewed (approved):
Looks awesome, thanks so much!
